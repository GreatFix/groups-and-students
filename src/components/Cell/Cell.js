import React from 'react'
import classes from './Cell.module.css'
import Label from '../Label/Label'
import Button from '../Button/Button'
import Sub from '../Sub/Sub'

const Cell = (props) => {
  return (
    <li className={`${classes.Cell} ${props.className}`}>
      <Label className={classes.Label} onClick={props.onClickLabel}>
        {props.children}
      </Label>
      {props.subContent && <Sub className={classes.Sub}>{props.subContent}</Sub>}

      {props.onClickEdit && (
        <Button className={classes.Edit} onClick={props.onClickEdit} color="info">
          EDIT
        </Button>
      )}
      {props.onClickDelete && (
        <Button className={classes.Delete} onClick={props.onClickDelete} color="danger">
          DELETE
        </Button>
      )}
    </li>
  )
}

export default Cell
