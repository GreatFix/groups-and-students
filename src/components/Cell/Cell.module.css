.Cell {
  list-style: none;
  display: flex;
  font-size: 1em;
}
.Index {
  flex-grow: 0.15;
  flex-basis: 15px;
  text-align: center;
}
.Label {
  flex-grow: 3;
  flex-basis: 50px;
  text-align: center;
}
.Sub {
  flex-grow: 1;
  flex-basis: 50px;
  text-align: center;
}
.Edit {
  flex-grow: 0.15;
  flex-basis: 15px;
}
.Delete {
  flex-grow: 0.15;
  flex-basis: 15px;
}
