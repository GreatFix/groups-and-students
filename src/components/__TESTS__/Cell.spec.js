import { screen, render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import userEvent from '@testing-library/user-event'
import Cell from '../Cell/Cell'

const TEST_ID = 1
const TEST_INDEX = 5
const TEST_SUBTEXT = 'TEST_SUBTEXT'
const TEST_CLASS = 'TEST_CLASS'
const TEST_TEXT = 'TEST_TEXT'

describe('Cell', () => {
  describe('renders with children and id', () => {
    const { container } = render(<Cell id={TEST_ID}>{TEST_TEXT}</Cell>)
    const cell = container.querySelector('.Cell')
    const simpleCell = cell.querySelector('.SimpleCell')

    test('renders SimpleCell', () => {
      expect(simpleCell).toHaveTextContent(TEST_TEXT)
    })
    test('render properly', () => {
      expect(cell).toMatchSnapshot()
    })
  })

  describe('renders with index', () => {
    const { container } = render(
      <Cell id={TEST_ID} index={TEST_INDEX}>
        {TEST_TEXT}
      </Cell>
    )
    const cell = container.querySelector('.Cell')
    const index = cell.querySelector('.Index')

    test('renders SimpleCell', () => {
      expect(index).toHaveTextContent(TEST_INDEX)
    })
    test('render properly', () => {
      expect(cell).toMatchSnapshot()
    })
  })

  describe('renders with subContent', () => {
    const { container } = render(
      <Cell id={TEST_ID} subContent={TEST_SUBTEXT}>
        {TEST_TEXT}
      </Cell>
    )
    const cell = container.querySelector('.Cell')
    const sub = cell.querySelector('.Sub')

    test('renders SimpleCell', () => {
      expect(sub).toHaveTextContent(TEST_SUBTEXT)
    })
    test('render properly', () => {
      expect(cell).toMatchSnapshot()
    })
  })

  describe('Clicks', () => {
    test('onClickDelete', () => {
      const handleClickDelete = jest.fn((id) => id)
      render(
        <Cell id={TEST_ID} onClickDelete={handleClickDelete}>
          {TEST_TEXT}
        </Cell>
      )
      userEvent.click(screen.getAllByAltText('delete')[0])
      expect(handleClickDelete).toHaveBeenCalledTimes(1)
      expect(handleClickDelete.mock.results[0].value).toBe(TEST_ID)
    })

    test('onClickEdit', () => {
      const handleClickEdit = jest.fn((id) => id)
      render(
        <Cell id={TEST_ID} onClickEdit={handleClickEdit}>
          {TEST_TEXT}
        </Cell>
      )
      userEvent.click(screen.getAllByAltText('edit')[0])
      expect(handleClickEdit).toHaveBeenCalledTimes(1)
      expect(handleClickEdit.mock.results[0].value).toBe(TEST_ID)
    })

    test('onClickLabel', () => {
      const handleClickLabel = jest.fn((id) => id)
      render(
        <Cell id={TEST_ID} onClickLabel={handleClickLabel}>
          {TEST_TEXT}
        </Cell>
      )
      userEvent.click(screen.getByText(TEST_TEXT))
      expect(handleClickLabel).toHaveBeenCalledTimes(1)
      expect(handleClickLabel.mock.results[0].value).toBe(TEST_ID)
    })
  })

  test('gets a class', () => {
    const { container } = render(
      <Cell id={TEST_ID} className={TEST_CLASS}>
        {TEST_TEXT}
      </Cell>
    )
    expect(container.querySelector('.Cell')).toHaveClass(TEST_CLASS)
  })
})
