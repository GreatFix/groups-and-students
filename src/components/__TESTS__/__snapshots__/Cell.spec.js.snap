// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell renders with children and id render properly 1`] = `
<li
  class="Cell "
>
  <p
    class="SimpleCell SimpleCell "
  >
    TEST_TEXT
  </p>
</li>
`;

exports[`Cell renders with index render properly 1`] = `
<li
  class="Cell "
>
  <p
    class="Sub Index"
  >
    5
  </p>
  <p
    class="SimpleCell SimpleCell "
  >
    TEST_TEXT
  </p>
</li>
`;

exports[`Cell renders with subContent render properly 1`] = `
<li
  class="Cell "
>
  <p
    class="SimpleCell SimpleCell "
  >
    TEST_TEXT
  </p>
  <p
    class="Sub Sub"
  >
    TEST_SUBTEXT
  </p>
</li>
`;
